x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/job/RestPostDataAction.java

```diff
@@ -27,11 +28,11 @@
 
     @Override
     public List<Route> routes() {
+        final String msg = "Posting data directly to anomaly detection jobs is deprecated, " +
+            "in a future major version it will be compulsory to use a datafeed";
         return List.of(
-            Route.builder(POST, BASE_PATH + "anomaly_detectors/{" + Job.ID + "}/_data")
-                .deprecated("Posting data directly to anomaly detection jobs is deprecated, " +
-                    "in a future major version it will be compulsory to use a datafeed", RestApiVersion.V_8
-                ).build()
+            Route.builder(POST, BASE_PATH + "anomaly_detectors/{" + Job.ID + "}/_data").deprecated(msg, RestApiVersion.V_8).build(),
+            Route.builder(POST, PRE_V7_BASE_PATH + "anomaly_detectors/{" + Job.ID + "}/_data").deprecated(msg, RestApiVersion.V_7).build()
```

Comment:
Is there a way to do a `replaces` AND have the additional deprecation message? It seems we are losing that the `_xpack` URL is old.

Reply:
I don't think it's possible to have two deprecation messages for the same endpoint.

It would be possible to have the `_ml` one say datafeeds will be compulsory in the future and the `_xpack` one say use the `_ml` one instead.

But I thought that "don't use this at all" was a more important thing to say than "use something different", especially as the different thing will then tell you not to use it at all.

It's not possible to use _just_ the post data endpoint - you have to use some other ML endpoints too.  So users would get the warnings about using `_xpack` for those other endpoints and hopefully realise that they needed to stop using `_xpack` everywhere.  Also, in 8.x the `_xpack` version only exists when the REST compatibility layer is used.  Presumably the user will test what happens when they stop using the REST compatibility layer before they stop using it in production.  So they'd find out that the `_xpack` version had disappeared at that point, and then it's a simple rename to get back to the deprecation message that the endpoint is going to be removed at some point.